# -*- coding: utf-8 -*-

# ArchToolkit - Archaeology Toolkit for QGIS
# Copyright (C) 2026 balguljang2
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
import os
import uuid
from qgis.PyQt import uic
from qgis.PyQt import QtWidgets
from qgis.PyQt.QtWidgets import QTableWidgetItem, QCheckBox, QWidget, QHBoxLayout, QFileDialog, QListWidgetItem
from qgis.PyQt.QtCore import Qt, QSize
from qgis.core import QgsProject, QgsVectorLayer
from qgis.PyQt.QtGui import QIcon
import processing
import tempfile
from .utils import push_message, restore_ui_focus
from .live_log_dialog import ensure_live_log_dialog

# Load the UI file
FORM_CLASS, _ = uic.loadUiType(os.path.join(
    os.path.dirname(__file__), 'dem_generator_dialog_base.ui'))

class DemGeneratorDialog(QtWidgets.QDialog, FORM_CLASS):
    # Map scale to recommended pixel size (meters)
    # Based on contour interval standards from National Geographic Information Institute
    SCALE_PIXEL_MAP = {
        '1:1,000 (ë“±ê³ ì„  1m)': 1.0,
        '1:2,500 (ë“±ê³ ì„  2m)': 2.0, 
        '1:5,000 (ë“±ê³ ì„  5m)': 5.0,
        '1:25,000 (ë“±ê³ ì„  10m)': 10.0,
        '1:50,000 (ë“±ê³ ì„  20m)': 20.0,
        'Custom (ì‚¬ìš©ì ì§€ì •)': None
    }
    
    # Interpolation methods with academic citations
    INTERPOLATION_METHODS = {
        'TIN - Linear (ì„ í˜•)': {
            'algorithm': 'qgis:tininterpolation',
            'method': 0,
            'desc': 'ğŸ’¡ ì‚¼ê°ë§ ê¸°ë°˜ ì„ í˜• ë³´ê°„. ë“±ê³ ì„  ë°ì´í„°ì— ì í•© [Delaunay, 1934]'
        },
        'TIN - Clough-Tocher (ê³¡ë©´)': {
            'algorithm': 'qgis:tininterpolation',
            'method': 1,
            'desc': 'ğŸ’¡ ì‚¼ê°ë§ ê¸°ë°˜ ê³¡ë©´ ë³´ê°„. ë¶€ë“œëŸ¬ìš´ ì§€í˜• í‘œí˜„ [Clough & Tocher, 1965]'
        },
        'IDW (ì—­ê±°ë¦¬ ê°€ì¤‘ì¹˜)': {
            'algorithm': 'qgis:idwinterpolation',
            'method': None,
            'desc': 'ğŸ’¡ í¬ì¸íŠ¸ ë°ì´í„°ì— ì í•©, ë“±ê³ ì„ ì—ëŠ” ë¹„ì¶”ì²œ [Shepard, 1968]'
        }
    }
    
    # DXF Layer definitions for Korean digital topographic maps
    # NOTE: Only essential contour lines default=True to avoid bridges/structures
    DXF_LAYER_INFO = {
        # --- í˜„í–‰ ìˆ˜ì¹˜ì§€í˜•ë„(ì¼ë°˜ì ìœ¼ë¡œ ë§ì´ ì“°ì´ëŠ” F*** ì½”ë“œ) ---
        'F0017110': {'name': 'ì£¼ê³¡ì„ ', 'desc': 'ê¸°ë³¸ ë“±ê³ ì„  (5m ê°„ê²©)', 'category': 'ë“±ê³ ì„ ', 'default': True},
        'F0017111': {'name': 'ê³„ê³¡ì„ ', 'desc': 'êµµì€ ë“±ê³ ì„  (25m ê°„ê²©)', 'category': 'ë“±ê³ ì„ ', 'default': True},
        'F0017112': {'name': 'ê°„ê³¡ì„ ', 'desc': 'ì™„ë§Œ ì§€í˜• íŒŒì„  (ì„ íƒì )', 'category': 'ë“±ê³ ì„ ', 'default': False},
        'F0017113': {'name': 'ì¡°ê³¡ì„ ', 'desc': 'ì•„ì£¼ ì™„ë§Œí•œ ì§€í˜• ì ì„ ', 'category': 'ë“±ê³ ì„ ', 'default': False},
        'F0017114': {'name': 'ì§€ì„±ì„ ', 'desc': 'ëŠ¥ì„ /ê³„ê³¡ ë³€í™”ì ', 'category': 'ì§€í˜•', 'default': False},
        'F0017115': {'name': 'ì§€ì„±ì„ (ì¶”ê°€)', 'desc': 'ì§€í˜• êµ´ê³¡ ë³´ì¡°', 'category': 'ì§€í˜•', 'default': False},
        'F0017120': {'name': 'ë“±ê³ ì„  ìˆ˜ì¹˜', 'desc': 'ë“±ê³ ì„  ìˆ«ì', 'category': 'í…ìŠ¤íŠ¸', 'default': False},
        'F0027111': {'name': 'í‘œê³ ì (ì§€í˜•)', 'desc': 'ìˆœìˆ˜ ì§€í˜• ë†’ì´ (ì‚°ì •ìƒ ë“±)', 'category': 'í¬ì¸íŠ¸', 'default': True},
        'F0027217': {'name': 'í‘œê³ ì (êµ¬ì¡°ë¬¼)', 'desc': 'âš ï¸ êµëŸ‰/êµ¬ì¡°ë¬¼ ë†’ì´ í¬í•¨ ì£¼ì˜!', 'category': 'í¬ì¸íŠ¸', 'default': False},
        'E0011111': {'name': 'í•˜ì²œì¤‘ì‹¬ì„ ', 'desc': 'í•˜ì²œ ë¬¼ê¸¸ (ê³ ë„ê°’ ì—†ì„ ìˆ˜ ìˆìŒ)', 'category': 'ìˆ˜ê³„', 'default': False},
        'E0011112': {'name': 'í•˜ì²œê²½ê³„ì„ ', 'desc': 'ê°•ë¬¼/ì§€ë©´ ê²½ê³„', 'category': 'ìˆ˜ê³„', 'default': False},
        'E0041311': {'name': 'í˜¸ìˆ˜/ì €ìˆ˜ì§€', 'desc': 'ìˆ˜ë©´ ê²½ê³„', 'category': 'ìˆ˜ê³„', 'default': False}
        ,
        # --- êµ¬(2000ë…„ëŒ€ ë“±) ìˆ˜ì¹˜ì§€í˜•ë„: ìˆ«ì ë ˆì´ì–´ ì½”ë“œ ---
        # ì£¼ë¡œ 71XX(ë“±ê³ ì„ ), 7217(í‘œê³ ì ), 73XX(ê¸°ì¤€ì /ìˆ˜ì¹˜) í˜•íƒœë¡œ ë“±ì¥í•©ë‹ˆë‹¤.
        # (ì˜ˆ) "Layer" IN ('7111','7114','7217' ...)
        "7111": {
            "name": "ì£¼ê³¡ì„ (êµ¬)",
            "desc": "êµ¬ ìˆ˜ì¹˜ì§€ë„(ìˆ«ì ì½”ë“œ) ë“±ê³ ì„ . (í˜„í–‰ ì˜ˆ: CAA002)",
            "category": "êµ¬ìˆ˜ì¹˜(ë“±ê³ ì„ )",
            "default": False,
        },
        "7114": {
            "name": "ê³„ê³¡ì„ (êµ¬)",
            "desc": "êµ¬ ìˆ˜ì¹˜ì§€ë„(ìˆ«ì ì½”ë“œ) ë“±ê³ ì„ (ê³„ê³¡ì„ ). (í˜„í–‰ ì˜ˆ: CAA001)",
            "category": "êµ¬ìˆ˜ì¹˜(ë“±ê³ ì„ )",
            "default": False,
        },
        "7217": {
            "name": "í‘œê³ ì (êµ¬)",
            "desc": "êµ¬ ìˆ˜ì¹˜ì§€ë„(ìˆ«ì ì½”ë“œ) í‘œê³ ì . (í˜„í–‰ ì˜ˆ: CA002)",
            "category": "êµ¬ìˆ˜ì¹˜(í‘œê³ ì )",
            "default": False,
        },
        "7132": {
            "name": "í‘œê³ ì ìˆ˜ì¹˜(êµ¬)",
            "desc": "êµ¬ ìˆ˜ì¹˜ì§€ë„(ìˆ«ì ì½”ë“œ) í‘œê³ ì  ìˆ˜ì¹˜(í…ìŠ¤íŠ¸/í‘œê¸°). DEM ë³´ê°„ì—ëŠ” ë³´í†µ ë¶ˆí•„ìš”",
            "category": "êµ¬ìˆ˜ì¹˜(í…ìŠ¤íŠ¸)",
            "default": False,
        },
        "2121": {
            "name": "í•´ì•ˆì„ (ìœ¡ì§€)(êµ¬)",
            "desc": "êµ¬ ìˆ˜ì¹˜ì§€ë„(ìˆ«ì ì½”ë“œ) í•´ì•ˆì„ (ìœ¡ì§€). í•´ì•ˆ/ìˆ˜ë©´ì„ 0m ê¸°ì¤€ìœ¼ë¡œ ì“°ê³  ì‹¶ì„ ë•Œë§Œ ì„ íƒ(ì£¼ì˜)",
            "category": "êµ¬ìˆ˜ì¹˜(í•´ì•ˆ)",
            "default": False,
        },
        "2122": {
            "name": "í•´ì•ˆì„ (ì„¬)(êµ¬)",
            "desc": "êµ¬ ìˆ˜ì¹˜ì§€ë„(ìˆ«ì ì½”ë“œ) í•´ì•ˆì„ (ì„¬). í•´ì•ˆ/ìˆ˜ë©´ ì²˜ë¦¬ë¥¼ ìœ„í•´ ì„ íƒí•  ìˆ˜ ìˆìŒ(ì£¼ì˜)",
            "category": "êµ¬ìˆ˜ì¹˜(í•´ì•ˆ)",
            "default": False,
        },
        # --- ì¼ë¶€ ë°ì´í„°ì…‹ì—ì„œ ì“°ì´ëŠ” ì˜ë¬¸ ì½”ë“œ(ì°¸ê³ ìš©) ---
        "CAA001": {
            "name": "ë“±ê³ ì„ (ê³„ê³¡ì„ )",
            "desc": "ì˜ë¬¸ ì½”ë“œ ë“±ê³ ì„ (ê³„ê³¡ì„ ). (êµ¬ ì½”ë“œ ì˜ˆ: 7114)",
            "category": "í˜„í–‰(ì˜ë¬¸ì½”ë“œ)",
            "default": False,
        },
        "CAA002": {
            "name": "ë“±ê³ ì„ (ì£¼ê³¡ì„ )",
            "desc": "ì˜ë¬¸ ì½”ë“œ ë“±ê³ ì„ (ì£¼ê³¡ì„ ). (êµ¬ ì½”ë“œ ì˜ˆ: 7111)",
            "category": "í˜„í–‰(ì˜ë¬¸ì½”ë“œ)",
            "default": False,
        },
        "CA002": {
            "name": "í‘œê³ ì ",
            "desc": "ì˜ë¬¸ ì½”ë“œ í‘œê³ ì . (êµ¬ ì½”ë“œ ì˜ˆ: 7217)",
            "category": "í˜„í–‰(ì˜ë¬¸ì½”ë“œ)",
            "default": False,
        },
        "CA0021": {
            "name": "í‘œê³ ì ìˆ˜ì¹˜",
            "desc": "ì˜ë¬¸ ì½”ë“œ í‘œê³ ì ìˆ˜ì¹˜(í‘œê¸°). DEM ë³´ê°„ì—ëŠ” ë³´í†µ ë¶ˆí•„ìš”",
            "category": "í˜„í–‰(ì˜ë¬¸ì½”ë“œ)",
            "default": False,
        },
    }

    DXF_LAYER_PRESETS = {
        "modern_f": {
            "label": "í˜„í–‰ ìˆ˜ì¹˜ì§€í˜•ë„ (F*** ë ˆì´ì–´)",
            "codes": ["F0017110", "F0017111", "F0027111"],
            "tooltip": (
                "í˜„í–‰ ìˆ˜ì¹˜ì§€í˜•ë„(DXF)ì—ì„œ ë§ì´ ë³´ì´ëŠ” F*** ë ˆì´ì–´ í”„ë¦¬ì…‹ì…ë‹ˆë‹¤.\n"
                "- ë“±ê³ ì„ (ì£¼ê³¡ì„ /ê³„ê³¡ì„ ) + í‘œê³ ì (ì§€í˜•)ë§Œ ê¸°ë³¸ ì„ íƒ\n"
                "- êµëŸ‰/êµ¬ì¡°ë¬¼ í‘œê³ ì (F0027217)ì€ ê¸°ë³¸ ì œì™¸"
            ),
        },
        "legacy_numeric": {
            "label": "êµ¬ ìˆ˜ì¹˜ì§€í˜•ë„ (ìˆ«ì ë ˆì´ì–´: 71XX/72XX/73XX)",
            "codes": ["7111", "7114", "7217", "2121", "2122"],
            "tooltip": (
                "êµ¬(2000ë…„ëŒ€ ë“±) ìˆ˜ì¹˜ì§€í˜•ë„ì—ì„œ ë ˆì´ì–´ ì´ë¦„ì´ ìˆ«ìë¡œ ë“¤ì–´ì˜¤ëŠ” ê²½ìš°ê°€ ìˆìŠµë‹ˆë‹¤.\n"
                "ì˜ˆ) \"Layer\" IN ('7111','7114','7217','2121','2122')\n"
                "- 71XX: ë“±ê³ ì„ , 7217: í‘œê³ ì \n"
                "- 2121/2122(í•´ì•ˆì„ )ì€ í•„ìš”í•  ë•Œë§Œ: í•´ì•ˆ/ìˆ˜ë©´ì„ 0m ê¸°ì¤€ìœ¼ë¡œ ê°•ì œí•  ìˆ˜ ìˆìŠµë‹ˆë‹¤(ì£¼ì˜)"
            ),
        },
        "modern_ca": {
            "label": "í˜„í–‰ ìˆ˜ì¹˜ì§€í˜•ë„ (ì˜ë¬¸ ì½”ë“œ: CA*/CAA*)",
            "codes": ["CAA001", "CAA002", "CA002"],
            "tooltip": (
                "ì¼ë¶€ ë°ì´í„°ì…‹ì€ ë“±ê³ ì„ /í‘œê³ ì  ë ˆì´ì–´ê°€ CAA001/CAA002/CA002 ê°™ì€ ì˜ë¬¸ ì½”ë“œë¡œ ë“¤ì–´ì˜µë‹ˆë‹¤.\n"
                "- ë“±ê³ ì„  + í‘œê³ ì ë§Œ ê¸°ë³¸ ì„ íƒ\n"
                "- (í•„ìš” ì‹œ) CA0021(í‘œê³ ì ìˆ˜ì¹˜)ì€ ìˆ˜ì¹˜ í‘œê¸°ë¼ DEM ë³´ê°„ì—ëŠ” ë³´í†µ ë¶ˆí•„ìš”"
            ),
        },
    }

    
    def __init__(self, iface, parent=None):
        super(DemGeneratorDialog, self).__init__(parent)
        self.setupUi(self)
        self.iface = iface
        self.loaded_dxf_layers = []
        
        # Initialize UI
        self.populate_layers()
        self.populate_scales()
        self.populate_interpolation_methods()
        self.setup_layer_table()
        self.setup_layer_presets()
        self.setup_layer_list()
        
        # Connect signals
        self.cmbScale.currentIndexChanged.connect(self.on_scale_changed)
        self.cmbInterpolation.currentIndexChanged.connect(self.on_interpolation_changed)
        self.btnLoadDxf.clicked.connect(self.load_dxf_file)
        self.btnSelectAll.clicked.connect(self.select_all_layers)
        self.btnDeselectAll.clicked.connect(self.deselect_all_layers)
        self.btnRefreshLayers.clicked.connect(self.populate_layers)
        self.btnRun.clicked.connect(self.run_process)
        self.btnClose.clicked.connect(self.reject)
        
        # Set button icon
        icon_path = os.path.join(os.path.dirname(os.path.dirname(__file__)), 'dem_icon.png')
        if os.path.exists(icon_path):
            self.btnRun.setIcon(QIcon(icon_path))
            self.btnRun.setIconSize(QSize(32, 32))
    
    def setup_layer_list(self):
        """Setup multi-select layer list with checkboxes"""
        self.listLayers.setSelectionMode(QtWidgets.QAbstractItemView.ExtendedSelection)
        self.listLayers.itemChanged.connect(self.on_layer_item_changed)
        self._updating_checkboxes = False
    
    def on_layer_item_changed(self, item):
        """When one checkbox is toggled, toggle all selected items too"""
        if self._updating_checkboxes:
            return
        
        self._updating_checkboxes = True
        new_state = item.checkState()
        
        # If this item is in selection, apply to all selected
        selected_items = self.listLayers.selectedItems()
        if item in selected_items:
            for sel_item in selected_items:
                sel_item.setCheckState(new_state)
        
        self._updating_checkboxes = False
    
    def populate_layers(self):
        """Populate layer list with vector layers (checkboxes)"""
        self.listLayers.clear()
        layers = QgsProject.instance().mapLayers().values()
        for layer in layers:
            if layer.type() == layer.VectorLayer:
                item = QListWidgetItem(layer.name())
                item.setData(Qt.UserRole, layer)
                item.setFlags(item.flags() | Qt.ItemIsUserCheckable)
                item.setCheckState(Qt.Unchecked)
                self.listLayers.addItem(item)
        
        # Auto-check layers containing 'DEMìš©' in name
        for i in range(self.listLayers.count()):
            item = self.listLayers.item(i)
            if 'DEMìš©' in item.text() or 'ë“±ê³ ì„ ' in item.text().lower():
                item.setCheckState(Qt.Checked)
    
    def setup_layer_table(self):
        """Setup the layer selection table with predefined DXF layers"""
        self.tblLayers.setColumnCount(4)
        self.tblLayers.setHorizontalHeaderLabels(['âœ“', 'ì½”ë“œ', 'ëª…ì¹­', 'ì„¤ëª…'])
        self.tblLayers.horizontalHeader().setStretchLastSection(True)
        self.tblLayers.setColumnWidth(0, 30)
        self.tblLayers.setColumnWidth(1, 80)
        self.tblLayers.setColumnWidth(2, 100)
        
        self.layer_checkboxes = {}
        row = 0
        self.tblLayers.setRowCount(len(self.DXF_LAYER_INFO))
        
        for layer_code, info in self.DXF_LAYER_INFO.items():
            checkbox = QCheckBox()
            checkbox.setChecked(info['default'])
            checkbox.setToolTip(f"{info['category']}: {info['desc']}")
            self.layer_checkboxes[layer_code] = checkbox
            
            widget = QWidget()
            layout = QHBoxLayout(widget)
            layout.addWidget(checkbox)
            layout.setAlignment(Qt.AlignCenter)
            layout.setContentsMargins(0, 0, 0, 0)
            self.tblLayers.setCellWidget(row, 0, widget)
            
            code_item = QTableWidgetItem(layer_code)
            code_item.setFlags(code_item.flags() & ~Qt.ItemIsEditable)
            self.tblLayers.setItem(row, 1, code_item)
            
            name_item = QTableWidgetItem(info['name'])
            name_item.setFlags(name_item.flags() & ~Qt.ItemIsEditable)
            self.tblLayers.setItem(row, 2, name_item)
            
            desc_item = QTableWidgetItem(info['desc'])
            desc_item.setFlags(desc_item.flags() & ~Qt.ItemIsEditable)
            self.tblLayers.setItem(row, 3, desc_item)
            
            row += 1

    def setup_layer_presets(self):
        """Add a compact 'layer preset' selector without changing the .ui file."""
        try:
            # horizontalLayout is defined in dem_generator_dialog_base.ui (row with SelectAll/Deselect/Load DXF).
            layout = getattr(self, "horizontalLayout", None)
            if layout is None:
                return

            self.lblLayerPreset = QtWidgets.QLabel("í”„ë¦¬ì…‹", self)
            self.cmbLayerPreset = QtWidgets.QComboBox(self)
            self.cmbLayerPreset.setMinimumWidth(220)

            # Add presets
            self.cmbLayerPreset.addItem("í”„ë¦¬ì…‹ ì„ íƒâ€¦", "")
            for key, item in self.DXF_LAYER_PRESETS.items():
                self.cmbLayerPreset.addItem(item.get("label", key), key)
                idx = self.cmbLayerPreset.count() - 1
                tip = item.get("tooltip", "")
                if tip:
                    self.cmbLayerPreset.setItemData(idx, tip, Qt.ToolTipRole)

            def _sync_tip():
                try:
                    self.cmbLayerPreset.setToolTip(
                        str(self.cmbLayerPreset.itemData(self.cmbLayerPreset.currentIndex(), Qt.ToolTipRole) or "")
                    )
                except Exception:
                    pass

            self.cmbLayerPreset.currentIndexChanged.connect(self.on_layer_preset_changed)
            self.cmbLayerPreset.currentIndexChanged.connect(_sync_tip)
            _sync_tip()

            # Insert after "ì„ íƒ í•´ì œ" (keeps the same row height)
            try:
                idx = int(layout.indexOf(self.btnDeselectAll))
                if idx >= 0:
                    layout.insertWidget(idx + 1, self.lblLayerPreset)
                    layout.insertWidget(idx + 2, self.cmbLayerPreset)
                else:
                    layout.insertWidget(0, self.lblLayerPreset)
                    layout.insertWidget(1, self.cmbLayerPreset)
            except Exception:
                try:
                    layout.insertWidget(0, self.lblLayerPreset)
                    layout.insertWidget(1, self.cmbLayerPreset)
                except Exception:
                    pass
        except Exception:
            pass

    def on_layer_preset_changed(self):
        key = ""
        try:
            key = str(self.cmbLayerPreset.currentData() or "")
        except Exception:
            key = ""

        if not key:
            return

        preset = self.DXF_LAYER_PRESETS.get(key) or {}
        codes = set(preset.get("codes") or [])
        if not codes:
            return

        # Apply: uncheck everything then check preset codes.
        for code, checkbox in (self.layer_checkboxes or {}).items():
            try:
                checkbox.setChecked(str(code) in codes)
            except Exception:
                continue
    
    def select_all_layers(self):
        for checkbox in self.layer_checkboxes.values():
            checkbox.setChecked(True)
    
    def deselect_all_layers(self):
        for checkbox in self.layer_checkboxes.values():
            checkbox.setChecked(False)
    
    def get_selected_layer_codes(self):
        selected = []
        for code, checkbox in self.layer_checkboxes.items():
            if checkbox.isChecked():
                selected.append(code)
        return selected
    
    def load_dxf_file(self):
        """Load multiple DXF files"""
        dxf_paths, _ = QFileDialog.getOpenFileNames(
            self,
            "DXF íŒŒì¼ ì„ íƒ (Ctrl+í´ë¦­ìœ¼ë¡œ ì—¬ëŸ¬ ê°œ ì„ íƒ)",
            "",
            "DXF Files (*.dxf);;All Files (*)"
        )
        
        if not dxf_paths:
            return
        
        selected_codes = self.get_selected_layer_codes()
        if not selected_codes:
            push_message(self.iface, "ì˜¤ë¥˜", "ìµœì†Œ í•˜ë‚˜ì˜ ë ˆì´ì–´ë¥¼ ì„ íƒí•´ì£¼ì„¸ìš”", level=2)
            restore_ui_focus(self)
            return
        
        query = '"Layer" IN (' + ','.join([f"'{code}'" for code in selected_codes]) + ')'
        
        total_features = 0
        loaded_count = 0
        
        for dxf_path in dxf_paths:
            try:
                layer_name = os.path.splitext(os.path.basename(dxf_path))[0] + "_DEMìš©"
                layer = QgsVectorLayer(dxf_path + "|layername=entities", layer_name, "ogr")
                
                if layer.isValid():
                    layer.setSubsetString(query)
                    QgsProject.instance().addMapLayer(layer)
                    self.loaded_dxf_layers.append(layer)
                    total_features += layer.featureCount()
                    loaded_count += 1
                    
            except Exception:
                push_message(self.iface, "ê²½ê³ ", f"{os.path.basename(dxf_path)} ë¡œë“œ ì‹¤íŒ¨", level=1)
        
        self.populate_layers()
        
        if loaded_count > 0:
            push_message(self.iface, "ì„±ê³µ", f"{loaded_count}ê°œ DXF ë¡œë“œ ì™„ë£Œ: ì´ {total_features}ê°œ í”¼ì²˜", level=0)
    
    def populate_scales(self):
        self.cmbScale.clear()
        for scale in self.SCALE_PIXEL_MAP.keys():
            self.cmbScale.addItem(scale)
        # Default to 1:5,000 (index 2)
        self.cmbScale.setCurrentIndex(2)
        self.on_scale_changed()
    
    def on_scale_changed(self):
        scale = self.cmbScale.currentText()
        recommended = self.SCALE_PIXEL_MAP.get(scale)
        
        if recommended is not None:
            self.spinPixelSize.setValue(recommended)
            self.lblRecommended.setText(f"(ê¶Œì¥: {recommended}m)")
        else:
            self.lblRecommended.setText("(ì§ì ‘ ì…ë ¥)")
    
    def populate_interpolation_methods(self):
        self.cmbInterpolation.clear()
        for method_name in self.INTERPOLATION_METHODS.keys():
            self.cmbInterpolation.addItem(method_name)
        self.on_interpolation_changed()
    
    def on_interpolation_changed(self):
        method_name = self.cmbInterpolation.currentText()
        method_info = self.INTERPOLATION_METHODS.get(method_name, {})
        desc = method_info.get('desc', '')
        self.lblInterpDesc.setText(desc)

    def get_selected_layers(self):
        """Get list of checked layers from the list widget"""
        selected_layers = []
        for i in range(self.listLayers.count()):
            item = self.listLayers.item(i)
            if item.checkState() == Qt.Checked:
                layer = item.data(Qt.UserRole)
                if layer:
                    selected_layers.append(layer)
        return selected_layers

    def run_process(self):
        """Run the DEM generation process (v0.7.2: Merge â†’ Filter â†’ Interpolate)"""
        selected_layers = self.get_selected_layers()
        output_path = self.fileOutput.filePath()
        pixel_size = self.spinPixelSize.value()
        
        if not selected_layers:
            push_message(self.iface, "ì˜¤ë¥˜", "ë ˆì´ì–´ë¥¼ ì²´í¬í•´ì£¼ì„¸ìš”", level=2)
            restore_ui_focus(self)
            return
        if not output_path:
            push_message(self.iface, "ì˜¤ë¥˜", "ì¶œë ¥ íŒŒì¼ ê²½ë¡œë¥¼ ì§€ì •í•´ì£¼ì„¸ìš”", level=2)
            restore_ui_focus(self)
            return

        # Live log window (non-modal) so users can see progress in real time.
        ensure_live_log_dialog(self.iface, owner=self, show=True, clear=True)

        method_name = self.cmbInterpolation.currentText()
        method_info = self.INTERPOLATION_METHODS.get(method_name, {})
        algorithm = method_info.get('algorithm', 'qgis:tininterpolation')
        method_param = method_info.get('method')
        
        # Build query for DXF layer filtering
        selected_codes = self.get_selected_layer_codes()
        
        # Auto-exclude bridge/structure elevation points
        BRIDGE_CODES = ['F0027217']  # êµëŸ‰/êµ¬ì¡°ë¬¼ í‘œê³ ì 
        excluded = len([c for c in selected_codes if c in BRIDGE_CODES])
        filtered_codes = [c for c in selected_codes if c not in BRIDGE_CODES]
        
        if filtered_codes:
            query = '"Layer" IN (' + ','.join([f"'{code}'" for code in filtered_codes]) + ')'
        else:
            query = None
        
        # Notify if bridge points were excluded
        if excluded > 0:
            push_message(self.iface, "ì•Œë¦¼", f"êµëŸ‰/êµ¬ì¡°ë¬¼ í‘œê³ ì  {excluded}ê°œ ìœ í˜• ìë™ ì œì™¸ë¨", level=0)
        
        push_message(self.iface, "ì²˜ë¦¬ ì¤‘", f"{len(selected_layers)}ê°œ ë ˆì´ì–´ ë³‘í•© ì¤‘...", level=0)
        self.hide()
        QtWidgets.QApplication.processEvents()
        
        try:
            temp_merged = None
            
            # Step 1: Merge all selected layers into one temp file
            if len(selected_layers) > 1:
                temp_merged = os.path.join(tempfile.gettempdir(), f'archtoolkit_merged_{uuid.uuid4().hex[:8]}.gpkg')
                processing.run("native:mergevectorlayers", {
                    'LAYERS': selected_layers,
                    'CRS': selected_layers[0].crs(),
                    'OUTPUT': temp_merged
                })
                merged_layer = QgsVectorLayer(temp_merged, "merged", "ogr")
            else:
                merged_layer = selected_layers[0]
            
            if not merged_layer or not merged_layer.isValid():
                push_message(self.iface, "ì˜¤ë¥˜", "ë ˆì´ì–´ ë³‘í•©ì— ì‹¤íŒ¨í–ˆìŠµë‹ˆë‹¤.", level=2)
                restore_ui_focus(self)
                return
            
            # Step 2: Apply query filter
            if query and merged_layer.fields().indexFromName('Layer') >= 0:
                merged_layer.setSubsetString(query)
            
            # Step 3: Find Z field
            z_field_idx = -1
            for fn in ['Z_COORD', 'z_coord', 'Elevation', 'ELEVATION', 'z_first']:
                idx = merged_layer.fields().indexFromName(fn)
                if idx >= 0:
                    z_field_idx = idx
                    break
            
            geom_type = merged_layer.geometryType()
            interp_type = 0 if geom_type == 0 else 1
            
            # Use source() for file-based layer
            source_path = merged_layer.source()
            
            if z_field_idx >= 0:
                interp_data = f'{source_path}::~::0::~::{z_field_idx}::~::{interp_type}'
            else:
                interp_data = f'{source_path}::~::1::~::0::~::{interp_type}'
            
            combined_extent = merged_layer.extent()


            
            params = {
                'INTERPOLATION_DATA': interp_data,
                'EXTENT': combined_extent,
                'PIXEL_SIZE': pixel_size,
                'OUTPUT': output_path
            }
            if method_param is not None:
                params['METHOD'] = method_param
            
            push_message(self.iface, "ì²˜ë¦¬ ì¤‘", "TIN ë³´ê°„ ì‹¤í–‰ ì¤‘...", level=0)
            QtWidgets.QApplication.processEvents()
            
            # Step 4: Run TIN interpolation
            result = processing.run(algorithm, params)
            
            # Add result to map
            if result and os.path.exists(output_path):
                self.iface.addRasterLayer(output_path, "ìƒì„±ëœ DEM")
                push_message(self.iface, "ì™„ë£Œ", f"DEM ìƒì„± ì™„ë£Œ! ({len(selected_layers)}ê°œ ë ˆì´ì–´ ë³‘í•©)", level=0)
                self.accept()
            else:
                push_message(self.iface, "ì˜¤ë¥˜", "DEMì´ ìƒì„±ë˜ì§€ ì•Šì•˜ìŠµë‹ˆë‹¤.", level=2)
                restore_ui_focus(self)
            
        except Exception as e:
            push_message(self.iface, "ì˜¤ë¥˜", f"ì²˜ë¦¬ ì¤‘ ì˜¤ë¥˜: {str(e)}", level=2)
            restore_ui_focus(self)
        finally:
            if temp_merged and os.path.exists(temp_merged):
                from .utils import cleanup_files
                cleanup_files([temp_merged])






